src/middlewared/middlewared/logger.py

```diff
@@ -69,14 +79,197 @@ def trace(self, message, *args, **kws):
+
+
+class TNSyslogHandler(logging.handlers.SysLogHandler):
     def __init__(
-        self, application_name: str, debug_level: typing.Optional[str]=None,
-        log_format: str='[%(asctime)s] (%(levelname)s) %(name)s.%(funcName)s():%(lineno)d - %(message)s'
+        self,
+        address: str = DEFAULT_SYSLOG_PATH,
+        pending_queue: deque | None = None
     ):
+        """
+        address - path to Unix socket (should be defined in syslog-ng config)
+
+        pending_queue - deque object that will be used for storing failed
+        LogRecords if syslog is currently down.
+
+        Note: maxlen should be set unless one wants to queue the log messages
+        without loss until syslog connection restored. This is probably
+        desireable for auditing, but not for general purposes (where it's
+        better to just specify a fallback handler).
+        """
+        self.pending_queue = pending_queue
+        self.fallback_handler = None
+        super().__init__(address, socktype=socket.SOCK_STREAM)
+
+    def drain_pending_queue(self) -> bool:
+        """
+        Attempt to emit any log records that have been queued up due to logging
+        failures to the syslog socket.
+
+        Returns:
+            True if successfully drained entire queue else False
+
+        Raises:
+            Should not raise exceptions
+        """
+        while self.pending_queue:
+            record = self.pending_queue.popleft()
+            try:
+                super().emit(record)
+            except Exception:
+                # Nope. Still dead. Put it back where we found it
+                self.pending_queue.appendleft(record)
+                return False
+
+        return True
+
+    def fallback(self, record: logging.LogRecord) -> None:
+        """
+        Fallback logging mechanism in case the syslog target is down.
+
+        In this case we emit the log record to the fallback handler and ignore
+        any errors.
+
+        Returns:
+            None
+
+        Raises:
+            Should not raise exceptions
+        """
+        if not self.fallback_handler:
+            return
+
+        try:
+            self.fallback_handler.emit(record)
+        except Exception:
+            pass
+
+    def emit(self, record: logging.LogRecord) -> None:
+        """
+        Emit a LogRecord to syslog. If this fails then add to pending queue and
+        emit via our fallback handler.
+        """
+
+        # First attempt to drain the pending queue
+        if not self.drain_pending_queue():
+            # Failed to drain our pending queue so add this record to the
+            # ever-growing deque
+            self.pending_queue.append(record)
+            self.fallback(record)
+            return
+
+        try:
+            super().emit(record)
+        except Exception:
+            # logging framework done broke. Queue up
+            # for drain on next auditd message handled
+            self.pending_queue.append(record)
```

Comment:
You might want to handle the edge-case when the queue is full. It'll raise an exception here. QueueFull or something, can't recall off top of head

Reply:
Nope, I misremembered. dequeu doesn't have that behavior.